instrumentation/opentelemetry-instrumentation-sklearn/src/opentelemetry/instrumentation/sklearn/__init__.py

```diff
@@ -0,0 +1,738 @@
+    model.  The assumption is that a machine learning model ``Pipeline`` (or
+    class descendent) is being instrumented with opentelemetry. Within a
+    ``Pipeline`` is some hierarchy of estimators and transformers.
+
+    The ``instrument_estimator`` method walks this hierarchy of estimators,
+    implementing each of the defined methods with its own span.
+
+    Certain estimators in the sklearn ecosystem contain other estimators as
+    instance attributes. Support for walking this embedded sub-hierarchy is
+    supported with ``recurse_attribs``. This argument is a dictionary
+    with classes as keys, and a list of attributes representing embedded
+    estimators as values. By default, ``recurse_attribs`` is empty.
+
+    Similar to Pipelines, there are also estimators which have class attributes
+    as a list of 2-tuples; for instance, the ``FeatureUnion`` and its attribute
+    ``transformer_list``. Instrumenting estimators like this is also
+    supported through the ``recurse_namedtuple_attribs`` argument. This
+    argument is a dictionary with classes as keys, and a list of attribute
+    names representing the namedtuple list(s). By default, the
+    ``recurse_namedtuple_attribs`` dictionary supports
+    ``Pipeline`` with ``steps``, and ``FeatureUnion`` with
+    ``transformer_list``.
+
+    Note that spans will not be generated for any child transformer whose
+    parent transformer has ``n_jobs`` parameter set to anything besides
+    ``None`` or ``1``.
+
+    Package instrumentation example:
+
+    .. code-block:: python
+
+        from opentelemetry.instrumentation.sklearn import SklearnInstrumentor
+
+        # instrument the sklearn library
+        SklearnInstrumentor().instrument()
+
+        # instrument additional libraries
+        packages = ["sklearn", "lightgbm", "xgboost"]
+        SklearnInstrumentor(packages=packages).instrument()
+
+
+    Model intrumentation example:
+
+    .. code-block:: python
+
+        from opentelemetry.instrumentation.sklearn import SklearnInstrumentor
+        from sklearn.datasets import load_iris
+        from sklearn.ensemble import RandomForestClassifier
+        from sklearn.model_selection import train_test_split
+        from sklearn.pipeline import Pipeline
+
+        X, y = load_iris(return_X_y=True)
+        X_train, X_test, y_train, y_test = train_test_split(X, y)
+
+        model = Pipeline(
+            [
+                ("class", RandomForestClassifier(n_estimators=10)),
+            ]
+        )
+
+        model.fit(X_train, y_train)
+
+        SklearnInstrumentor().instrument_estimator(model)
+
+    Args:
+        methods (list): A list of method names on which to instrument a span.
+          This list of methods will be checked on all estimators in the model
+          hierarchy. Used in package and model instrumentation
+        recurse_attribs (dict): A dictionary of ``BaseEstimator``-derived
```

Comment:
Just wondering why we give users the option to provide their own `recurse_attribs` and `recurse_namedtuple_attribs` to override only certain estimators and methods? The signature for this already quite complicated, and we don't have this kind of mechanism for any other instrumentation. Perhaps for the original PR implementation, we should just have the default attributes overridden?

Reply:
If this is a desired functionality of sklearn users however, I don't mind having it here.

Reply:
This is necessary due to the extensibility of the sklearn API. If a user has extended sklearn's API into their own classes AND wish to use the model instrumentation, then they have to provide this configuration. If they define metaestimators like `Pipeline` and `FeatureUnion`, then they must declare the underlying estimator attributes in order for model instrumentation to follow the hierarchy properly.

Reply:
>Perhaps for the original PR implementation, we should just have the default attributes overridden?

Maybe I misread this the first time but you're suggesting just removing them from the signature and controlling them using the module variables?